src/goto-instrument/havoc_utils.cpp

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************\
+
+Module: Utilities for building havoc code for expressions.
+
+Author: Saswat Padhi, saspadhi@amazon.com
+
+Date: July 2021
+
+\*******************************************************************/
+
+/// \file
+/// Utilities for building havoc code for expressions
+
+#include "havoc_utils.h"
+
+#include <util/arith_tools.h>
+#include <util/c_types.h>
+#include <util/expr_util.h>
+#include <util/pointer_expr.h>
+
+class havoc_utils_is_constantt : public is_constantt
+{
+public:
+  explicit havoc_utils_is_constantt(const modifiest &mod) : modifies(mod)
+  {
+  }
+  bool is_constant(const exprt &expr) const override
+  {
+    // Besides the "usual" constants (checked in is_constantt::is_constant),
+    // we also treat unmodified symbols as constants
+    if(expr.id() == ID_symbol && modifies.find(expr) == modifies.end())
+      return true;
+
+    return is_constantt::is_constant(expr);
+  }
+
+protected:
+  const modifiest &modifies;
+};
+
+void append_safe_havoc_code_for_expr(
+  const source_locationt source_location,
+  const exprt &expr,
+  goto_programt &dest,
+  std::function<void()> havoc_code_impl)
+{
+  goto_programt skip_program;
+  goto_programt::targett skip_target =
+    skip_program.add(goto_programt::make_skip(source_location));
+
+  if(expr.id() == ID_dereference)
+  {
+    // check for validity of underlying pointer and skip havocing if invalid
+    const auto condition = not_exprt(w_ok_exprt(
+      to_dereference_expr(expr).pointer(),
+      from_integer(0, unsigned_int_type())));
+    dest.add(goto_programt::make_goto(skip_target, condition, source_location));
+  }
+
+  havoc_code_impl();
+
+  if(expr.id() == ID_dereference)
+    dest.destructive_append(skip_program);
+}
+
+void append_object_havoc_code_for_expr(
+  const source_locationt source_location,
+  const exprt &expr,
+  goto_programt &dest)
+{
+  append_safe_havoc_code_for_expr(source_location, expr, dest, [&]() {
+    codet havoc(ID_havoc_object);
+    havoc.add_source_location() = source_location;
+    havoc.add_to_operands(expr);
+
+    dest.add(goto_programt::make_other(havoc, source_location));
+  });
+}
+
+void append_scalar_havoc_code_for_expr(
+  const source_locationt source_location,
+  const exprt &expr,
+  goto_programt &dest)
+{
+  append_safe_havoc_code_for_expr(source_location, expr, dest, [&]() {
+    side_effect_expr_nondett rhs(expr.type(), source_location);
```

Comment:
Again, if `expr.source_location() == source_location`, we wouldn't need the extra param.

Reply:
For `expr` that comes from an annotation (like assigns clause) this would be fine, but for the auto-computed modifies set, `expr` would have arbitrary source locations within the loop body. This extra source location is currently used to override those locations and just show the location of the loop head for the auto-computed modifies set.

Reply:
Thanks for the clarification!